### ISSUE TITLE
queryserver improvements to avoid overload

### ISSUE THREAD

**Issue by demmer** [vitessio/vitess#3718]
Recently we ran into an issue in load testing where our application overwhelmed vtqueryserver with incoming read requests, including several particularly expensive queries in mysql. This ended up blocking application requests for a long time and tied up resources since the connections in the query pool were tied up with incoming requests.

Although in our case this happened using vtqueryserver, some similar issues would occur in "full" vitess. I think the following three features would help address this issue and I plan to follow up with PRs for all of them:

1. Add a configurable wait timeout to get a connection from the query pool: #3717
2. Add a limit on the number of queries that can be blocked waiting for a connection from either the query pool or the transaction pool.
3. In the mysql protocol server (used both in vtgate and vtqueryserver), use a cancellable Context instead of Context.Background, and cancel any in-flight queries if the client connection goes away.

**Comment by demmer** [comment]
@alainjobart Looking at the code I don't see an easy way that we could accomplish number 3 above without a major rewrite of how the server code works. More specifically -- the ComQuery handler blocks  when executing the query which means that nothing is checking the client socket for an EOF or other interrupt. As a result we don't have a great signal to cancel an in-flight context and cause the query to be killed.

IMO this would be a good feature to have both for the case when the client app went away and for interactive use cases where a user hits Control-C to the mysql client.

Can you verify that my understanding is correct and if so, whether (and how) you see us accomplishing the intended goal?

**Comment by alainjobart** [comment]
Yes I agree with your analysis. The ComQuery handler definitely blocks until there is a response from the Vitess (in vtgate) / MySQL (in vtqueryserver) side. So if the client goes away in the middle of the query, we still finish that query, then try to write the result, fail, and then clean up.

I think the only way to fix this is to de-correlate the processing of the client socket events from the actual execution of the queries. Right now, we have the following code, for each accepted connection:
* finish handshake
* loop:
  * read packet
  * call callback synchronously (no way to interrupt the call)
  * write response

(and this code is complicated already, because we try to re-use buffers between the reading and the writing. That part is probably too much of an optimization, and should be simplified / made more robust).

I think this should be changed into at least two routines, which share a single 'state' structure. The 'state' structure would contain the current state of the connection (idle, processing request, writing response), and if processing a request, also have a context and a cancelFunc for that request:

* the 'read' routine, a forever loop:
  * read a packet. Would return EOF if the connection was closed.
  * lock 'state'. Then:
    * if EOF, and a query is in flight, cancel its context.
    * if EOF, and no query is in flight, we're done.
    * if query, and if a query is already in flight, it's an error, close everything.
    * if query, and if no query is in flight, change state to 'processing query', create a context/cancelFunc for it, create a go routine and call it, with the context.
* the go routine to process a query would call the callback for it, passing in the context, then wait for the response. When it gets it, lock the state again, switch the state to 'writing response', unlock the state, write the response, then lock the state, switch state back to 'idle', unlock the state.

That way, EOF can be read from the socket at the same time as the query is on-going, and cancel its context when that happens. Effectively forwarding the 'close' event.

Note the processing API (the Handler interface) would just get an extra context.Context as first parameter of ComQuery(...) and the implementations can just use it for their queries.

Side note 1: as I mentioned earlier, the buffer processing in go/mysql is trying to be too optimized, and re-use buffers as much as possible, at a high complexity cost in the code. @sougou  and I have been talking about simplifying this for a bit, by writing a small optimized memory allocator layer, that uses various pools of memory buffers. That would greatly simplify the code everywhere (no more ephemeral packets and so on) while keeping good performance. That refactoring would probably need to be done first, otherwise the buffer management complexity would make this change look way too complex.

Side note 2: we don't have an idle timeout in the server side: of a client opens a connection, and keeps it open forever, we will forever listen to it. MySQL has an idle timeout, we probably should have one too. One way to achieve this: keep a list / map of all connections. Every minute or so, go through them all, and see which ones have been idle for more than idle timeout, and close them.

Side note 3: we don't have a maximum number of in-flight connections, and we probably should. one way to achieve this: keep a list / map of all connections. When we accept, refuse to accept if len(list) is too big. Or we could finish authentication phase, and then check connection count (which could then be per user).

**Comment by danieltahara** [comment]
Thinking about this a bit and comparing to the proxy I'm working on replacing:

Obviously the buffer re-use is a problem that will need to be fixed, but architecturally this doesn't strike me as needing that much state -- our existing proxy at Dropbox just has the main thread sit and collect client packets and pass them over a channel (well, it's in python) to an executor that just loops until its iterator ends. That's pretty easily modelable in golang, assuming we can fix the buffer reuse problem.

**Comment by demmer** [comment]
Re Side note 2, @rafael just added this in #3911 
Re Side note 3, I think that's a good idea too. We should add that for safety.

Re the overall approach, @rafael and I also sketched out something which might be even simpler -- in the implementation of ComQuery, once the query packet arrives, create a new context and channel, and run a new goroutine to execute the query using this context. When done running the query, this would send a completion notice to the channel.

Then the original goroutine could select on the connection socket or the completion channel. If it gets either EOF (or a packet) from the socket, then it knows the client went away (or sent a bogus protocol message). But in the normal case it would get the completion record from the channel and could send the response.

So although I agree the buffer reuse is probably over-optimized, it seems like we can handle the cancellation using this approach without needing a big rewrite of the buffering model.

**Comment by danieltahara** [comment]
+1 to the goroutine per request approach, since that's basically what an rpc server would do anyway (maybe just throw it under a semaphore per all the pools).
